exchanges/graphcache/src/offlineExchange.ts

```diff
@@ -0,0 +1,141 @@
+import { pipe, filter } from 'wonka';
+import { print, SelectionNode } from 'graphql';
+
+import {
+  Operation,
+  GraphQLRequest,
+  Exchange,
+  CombinedError,
+  createRequest,
+} from '@urql/core';
+
+import {
+  getMainOperation,
+  getFragments,
+  isInlineFragment,
+  isFieldNode,
+  shouldInclude,
+  getSelectionSet,
+  getName,
+} from './ast';
+
+import { makeDict } from './helpers/dict';
+import { OptimisticMutationConfig, Variables } from './types';
+import { cacheExchange, CacheExchangeOpts } from './cacheExchange';
+
+/** Determines whether a given query contains an optimistic mutation field */
+const isOptimisticMutation = (
+  config: OptimisticMutationConfig,
+  operation: Operation
+) => {
+  if (operation.operationName !== 'mutation') {
+    return false;
+  }
+
+  const vars: Variables = operation.variables || makeDict();
+  const fragments = getFragments(operation.query);
+  const selections = [...getSelectionSet(getMainOperation(operation.query))];
+
+  let field: void | SelectionNode;
+  while ((field = selections.pop())) {
+    if (!shouldInclude(field, vars)) {
+      continue;
+    } else if (!isFieldNode(field)) {
+      const fragmentNode = !isInlineFragment(field)
+        ? fragments[getName(field)]
+        : field;
+      if (fragmentNode) selections.push(...getSelectionSet(fragmentNode));
+    } else if (config[getName(field)]) {
+      return true;
+    }
+  }
+
+  return false;
+};
+
+const isOfflineError = (error: undefined | CombinedError) =>
+  error &&
+  error.networkError &&
+  !error.response &&
+  ((typeof navigator !== 'undefined' && navigator.onLine === false) ||
+    /request failed|failed to fetch|network\s?error/i.test(
+      error.networkError.message
+    ));
+
+export const offlineExchange = (opts: CacheExchangeOpts): Exchange => ({
+  forward: outerForward,
+  client,
+  dispatchDebug,
+}) => {
+  const { storage } = opts;
+  let forward = outerForward;
+
+  if (
+    storage &&
+    storage.onOnline &&
+    storage.readMetadata &&
+    storage.writeMetadata
+  ) {
+    const optimisticMutations = opts.optimistic || {};
```

Comment:
One of the things I'm wondering is how we can tell our wrapped `cacheExchange` not to clear the optimistic layer when we get the "you are offline" error